dali/operators/reader/loader/tar_utils.cc

```diff
@@ -0,0 +1,197 @@
+// Copyright (c) 2021, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/operators/reader/loader/tar_utils.h"
+#include <libtar.h>
+#include <algorithm>
+#include <cstdarg>
+#include <cstdlib>
+#include <list>
+#include <string>
+#include <tuple>
+#include <utility>
+#include "dali/core/error_handling.h"
+#include "dali/core/math_util.h"
+#include "dali/core/util.h"
+
+namespace dali {
+namespace detail {
+
+namespace {
+
+constexpr uint64_t kBlockSize = T_BLOCKSIZE;
+static_assert(is_pow2(kBlockSize),
+              "The implementation assumes that the block size is a power of 2");
+
+constexpr uint64_t kEmptyEofBlocks = 2;
+constexpr uint64_t kTarArchiveBufferInitSize = 1;
+
+std::mutex instances_mutex;
+std::list<std::vector<TarArchive*>> instances_registry = {
+    std::vector<TarArchive*>(kTarArchiveBufferInitSize)};
+TarArchive** instances = instances_registry.back().data();
+
+int Register(TarArchive* archive) {
+  std::lock_guard<std::mutex> instances_lock(instances_mutex);
+  for (auto& instances_entry : instances_registry.back()) {
+    if (instances_entry != nullptr) {
+      continue;
+    }
+    instances_entry = archive;
+    return &instances_entry - instances;
+  }
+  std::vector<TarArchive*>& old = instances_registry.back();
+  instances_registry.emplace_back();
+  std::vector<TarArchive*>& curr = instances_registry.back();
+  curr.reserve(old.size() * 2);
+  curr = old;
+  curr.push_back(archive);
+  curr.resize(old.size() * 2, nullptr);
+  instances = curr.data();
+  return old.size();
+}
+
+inline void Unregister(int instance_handle) {
+  instances[instance_handle] = nullptr;
+}
+
+}  // namespace
+
+TarArchive::TarArchive(std::unique_ptr<FileStream> stream)
+    : stream(std::move(stream)), archiveoffset(0), instance_handle(Register(this)), eof(false) {
+  this->stream->Seek(0);
+  ParseHeader();
+}
+
+TarArchive::TarArchive(TarArchive&& other) {
+  *this = static_cast<TarArchive&&>(other);
+}
+
+TarArchive::~TarArchive() {
+  Unregister(instance_handle);
+}
+
+TarArchive& TarArchive::operator=(TarArchive&& other) {
+  stream = std::move(other.stream);
+  eof = other.eof;
+  filename = other.filename;
+  filesize = other.filesize;
+  readoffset = other.readoffset;
+  archiveoffset = other.archiveoffset;
+
+  std::lock_guard<std::mutex> instances_lock(instances_mutex);
+  instances[instance_handle] = nullptr;
```

Comment:
```suggestion
  instances[instance_handle] = this;
```
?

Reply:
cause it gets moved and after getting moved this spot is going to be empty